File: nsr_api_example.php
<?php

//////////////////////////////////////////////////////
// Test submission to NSR API
//
// Adjust parameters to test, especially $mode
// Requires: nsr_testfile.dsv
// Compatible with NSR version: 2.3+
//////////////////////////////////////////////////////

// Load functions
require("php_utilities/functions.inc");
require("php_utilities/status_codes.inc.php");

/////////////////////
// Parameters
/////////////////////

//
// API parameters
//

// Path and name of file containing input names and political divisions
$DATADIR = "data/user/";
$inputfilename = "nsr_testfile.csv";

// Number of lines to import
// Start with a very small number for testing
// Set to large number to impart entire file
$lines = 1000;

// api base url 
$base_url = "https://bien.nceas.ucsb.edu/nsr/nsr_wsb.php";	# Prod
//$base_url = "https://bien.nceas.ucsb.edu/nsrdev/nsr_wsb.php"; 	# Dev
$base_url = "https://bien.nceas.ucsb.edu/nsrdev/nsr_wsb.php"; 	# Dev.new

//
// NSR options
//

// Processing mode
//	Options: resolve*|parse|meta
// 	E.g., $mode="parse"
$mode="resolve";	# Resolve native status
// $mode="meta";		# Return metadata on NSR & sources
// $mode="sources";		# List NSR sources
// $mode="citations";		# Return citations for NSR & sources

//
// Display options
// 
// * Turn on/off what is echoed to terminal
// * Raw data always displayed
//

$disp_data_array=false;		// Echo raw data as array
$disp_combined_array=false;	// Echo combined options+data array
$disp_opts_array=false;		// Echo NSR options as array
$disp_opts=true;			// Echo NSR options
$disp_json_data=false;		// Echo the options + raw data JSON POST data
$disp_results_json=true;	// Echo results as array
$disp_results_array=false;	// Echo results as array
$disp_results_csv=true;		// Echo results as CSV text, for pasting to Excel

/////////////////////
// Main
/////////////////////

///////////////////////////////
// Make options array
///////////////////////////////

$opts_arr = array(
"mode"=>$mode
);

///////////////////////////////
// Prepare the json request body
///////////////////////////////

if ($mode=="resolve") {
	// Import csv data and convert to data array
	$data_arr = array_map('str_getcsv', file($DATADIR.$inputfilename));

	# Get subset
	$data_arr = array_slice($data_arr, 0, $lines);

	// Echo the raw data
	if ( $mode=="resolve" || $mode=="" ) {
		// Echo raw data
		echo "The raw data:\r\n";
		foreach($data_arr as $row) {
			foreach($row as $key => $value) echo "$value\t"; echo "\r\n";
		}
		echo "\r\n";

		if ($disp_data_array) {
			echo "The raw data as array:\r\n";
			var_dump($data_arr);
			echo "\r\n";
		}
	}
	
	$input_json = json_encode(array('opts' => $opts_arr, 'data' => $data_arr));	
} else {
	//$data_arr=array();
	$input_json = json_encode(array('opts' => $opts_arr));	
}

// Echo the JSON body
echo "The JSON input for the API:\r\n";
echo $input_json . "\r\n\r\n";

///////////////////////////////
// Decompose the JSON into opt 
// and data (for display)
///////////////////////////////

$input_array = json_decode($input_json, true);

if ($disp_combined_array) {
	echo "The combined array:\r\n";
	var_dump($input_array);
	echo "\r\n";
}

$opts = $input_array['opts'];
if ($disp_opts_array) {
	echo "Options array:\r\n";
	var_dump($opts);
	echo "\r\n";
}

if ($disp_opts) {
	// Convert booleans to text for display
	$mode_disp = isset($opts['mode']) ? $mode : "resolve";
	
	// Echo the options
	echo "NSR options:\r\n";
	echo "  mode: " . $mode_disp . "\r\n";
	echo "\r\n";
}

if ($disp_json_data) {
	// Echo the final JSON post data
	echo "API input (options + raw data converted to JSON):\r\n";
	echo $input_json . "\r\n\r\n";
}

///////////////////////////////
// Call the API
///////////////////////////////

$url = $base_url;    
$content = $input_json;

// Initialize curl & set options
$ch = curl_init($url);	
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);	// Return response (CRITICAL!)
curl_setopt($ch, CURLOPT_POST, 1);	// POST request
curl_setopt($ch, CURLOPT_POSTFIELDS, $input_json);	// Attach the encoded JSON
curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-Type: application/json')); 

// Execute the curl API call
$response = curl_exec($ch);

// Check status of the response and echo if error
$status = curl_getinfo($ch, CURLINFO_HTTP_CODE);

// if ( $status != 201 && $status != 200 ) {
//     die("Error: call to URL $url failed with status $status, response $response, curl_error " . curl_error($ch) . ", curl_errno " . curl_errno($ch) . "\r\n");
// }
if ( $status != 201 && $status != 200 ) {
	$status_msg = $http_status_codes[$status];
    die("Error: call to URL $url failed with status $status \r\n$status_msg \r\nDetails: $response \r\n");
}

// Close curl
curl_close($ch);

$results_json = $response;
$results = json_decode($results_json, true);

// Echo the response content
echo "API results (JSON)\r\n";
echo $results_json;
echo "\r\n\r\n";

echo "API results as CSV:\r\n";
if ( $mode=="parse" || $mode=="" ) {
	foreach($results as $result) {
		$line = implode(",", array_slice($result, 0)) . "\r\n";
		print_r($line);
	}
} else {
	$table = array_to_csv($results,",");
	echo $table;
}
echo "\n";

?>

File: nsr_api_example.R
###############################################
# R NSR API Example
###############################################

rm(list=ls())

#################################
# Parameters & libraries
#################################

##################
# Base URL
##################

url = "https://bien.nceas.ucsb.edu/nsr/nsr_wsb.php"

##################
# Libraries
##################

library(httr)		# API requests
library(jsonlite) # JSON coding/decoding

##################
# Test data
##################

# Test data
# For details on how to organize an NSR input file, see:
# http://bien.nceas.ucsb.edu/bien/tools/nsr/batch-mode/
example_file <- "https://bien.nceas.ucsb.edu/bien/wp-content/uploads/2023/06/nsr_testfile.csv"

# Set TRUE to use sample of 10 rows of data, FALSE to use all
data.use.sample <- FALSE

##################
# Misc parameters
##################

# API variables to clear before each API call
# Avoids spillover between calls
api_vars <- c("mode")

# Response variables to clear before each API call
# Avoids confusion with previous results if API call fails
response_vars <- c("request_json", "response_json", "response")

#################################
# Functions
#################################

make_request_json <- function( 
	mode,				# API mode; required
	data=NULL 	# Raw data; required if mode=='resolve'
	) {
	######################################
	# Accepts: options parameters and (optionally) data
	# Returns: formatted JSON api request
	######################################
	
	# Set defaults if applicable
	if ( mode=="resolve"  )  {
		# Convert raw data to JSON
		if (is.null(data) ) stop("ERROR: mode 'resolve' requires data!\n")
		data_json <- jsonlite::toJSON(unname(data))
	} 
	
	opts <- data.frame(mode = mode)
	opts_json <-  jsonlite::toJSON(opts)
	opts_json <- gsub('\\[','',opts_json)
	opts_json <- gsub('\\]','',opts_json)
	
	# Combine the options and data into single JSON object
	if ( mode=="resolve" ) {
		input_json <- paste0('{"opts":', opts_json, ',"data":', data_json, '}' )
	} else {
		input_json <- paste0('{"opts":', opts_json, '}' )
	}
	
	return(input_json)
}

send_request_json <- function( url, request_json ) {
	###################################
	# Accepts: API url + JSON options & data
	# Sends: POST request to url, with JSON
	#		attached to body
	# Returns: JSON response
	###################################
	if ( is.null(url) || is.na(url) ) {
		stop("ERROR: parameter 'url' missing (function send_request_json()'\n")
	}
	if ( is.null(request_json) || is.na(request_json) ) {
		stop("ERROR: parameter 'json_body' missing (function send_request_json()'\n")
	}
	
	response_json <- POST(url = url,
	                  add_headers('Content-Type' = 'application/json'),
	                  add_headers('Accept' = 'application/json'),
	                  add_headers('charset' = 'UTF-8'),
	                  body = request_json,
	                  encode = "json"
	                  )

	return(response_json)
}

decode_response_json <- function( response_json, mode ) {
	###################################
	# Converts resonse json to data frame
	# Also decodes weird format of NSR JSON
	###################################
	
	response_raw <- fromJSON( rawToChar( response_json$content ) ) 
	response <- as.data.frame(response_raw)
	
	if ( mode=="resolve") {
		# Extra transformations needed due to weird NSR response data format
		col.names <- response$id
		response <- as.data.frame(t(response[,-1]))
		colnames(response) <- col.names
		
		# Convert f'ing factors to character
		factor_columns <- sapply(response, is.factor)
		response[factor_columns] <- lapply(
			response[factor_columns], as.character
			)
		# Convert numeric columns
		response$isIntroduced <- as.integer(response$isIntroduced)
		response$isCultivatedNSR <- as.integer(response$isCultivatedNSR)
		
		# Reset row numbers
		row.names(response) <- 1:nrow(response)
	} else {
		# Unnest
		response <- response[[1]]
	}
	
	return( response )
}

get_request <- function(
		url, 					# Required
		mode,				# Required
		data=NULL 	# Raw data; required if mode=='resolve)
	) {
	######################################
	# Accepts: options parameters and (optionally) data
	#		required for API request
	# Sends: POST request to API
	# Returns: response formatted as data frame
	# 
	# Meta-function which combine functions 
	# make_request_json, send_request_json & 
	# decode_response_json. See component 
	# functions for details
	######################################
	if ( is.null(url) || is.na(url) ) {
		stop("ERROR: parameter 'url' missing (function get_request()'\n")
	}
	if ( is.null(mode) || is.na(mode) ) {
		stop("ERROR: parameter 'mode' missing (function get_request()'\n")
	}

	request_json <- make_request_json(
		mode=mode,
		data= data 
		)
	response_json <- send_request_json( url, request_json )
	response_df <- decode_response_json( response_json, mode=mode )
	
	# if ( ncol(response_df)==1 ) {
		# colnames(response_df) <- "error"
		# response_df$http_status <- response_json$status
	# }

	return( response_df )
}

specify_decimal <- function(x, k) {
	# Set fixed number of decimals
	if ( is.na(x) || is.null(x) ) {
		x.formatted <- x
	} else {
		x.formatted <- format(round(x, k), nsmall=k)
	}
	return( x.formatted )
}

#################################
# Selected metadata checks
#
# Available metadata calls: 
# "meta", "sources", "citations", "checklist_countries", "country_checklists"
# print(response) to see the complete results of each call
#################################

# Application version and database version
mode <- "meta"		
rm( list = Filter( exists, response_vars ) )
# request_json <- make_request_json(mode=mode)
# response_json <- send_request_json( url, request_json )
# response <- decode_response_json(response_json)
response <- get_request(url=url, mode=mode)
db_version  <- response$db_version
db_date   <- response$build_date
if( "app_version" %in% colnames(response) ) {
	app_version  <- response$app_version
} else {
	app_version  <- response$code_version
}

# Available sources
mode <- "sources"		
rm( list = Filter( exists, response_vars ) )
response <- get_request(url=url, mode=mode)
source.details <- response

# App and source citations
mode <- "citations"		
rm( list = Filter( exists, response_vars ) )
citations <- get_request(url=url, mode=mode)

# Display results
cat("NSR version: ", app_version, "\n")
cat("DB version: ", db_version, " (", db_date, ")\n")
cat("Checklists:\n")
print(source.details, row.names=FALSE)
cat("Citations:")
print(citations, row.names=FALSE)

#################################
# Load test data
#################################

data <- read.csv(example_file, header=TRUE)
if (data.use.sample==TRUE) data <- head(data, 10) # Just a sample
cat("Raw data:\n")
print(data)

#################################
# Example 1: Resolve mode
#################################

# Clear existing variables
suppressWarnings( rm( list = Filter( exists, c(response_vars, api_vars ) ) ) )

# Set options
mode <- "resolve"			# Processing mode

response <- get_request(url=url, mode=mode, data=data	)
if ( colnames(response)[1]=="error" ) {
	print( response )
} else {
	print(response)
	
	# More compact display:
	print(response[ , c("family", "species", "country", "state_province", "native_status", "native_status_reason", "native_status_sources")])
}

#################################
# Example 2: Country checklists
#################################

mode <- "country_checklists"		
rm( list = Filter( exists, response_vars ) )
country.checklists <- get_request(url=url, mode=mode)
country.checklists <- country.checklists[ order(country.checklists$gid_0), ]
print(country.checklists)

#################################
# Example 3: Checklist countries
#################################

mode <- "checklist_countries"		
rm( list = Filter( exists, response_vars ) )
checklist.countries <- get_request(url=url, mode=mode)
print(checklist.countries)
